### PR TITLE
[examples] use prompt api to re-implement few-shot methods

### PR DESCRIPTION
**Pull request by LemonNoel**
### PR types

Others

### PR changes

Others

### Description

- Add PET, P-Tuning and EFL examples based on Prompt API

P-Tuning Framework
image

EFL Framework
截屏2022-11-28 17 41 09

### CODE REVIEW

File: examples/few_shot/p-tuning/README.md [hunk 1]
```diff
@@ -1,113 +1,82 @@
...
+
 PaddleNLP 内置了 FewCLUE 数据集，可以直接用来进行 P-tuning 策略训练、评估、预测，并生成 FewCLUE 榜单的提交结果，参与 FewCLUE 竞赛。
 
+### 代码结构及说明
+```
+|—— train.py # P-tuning 算法训练、评估主程序入口
+|—— data.py  # 数据集构造、数据增强等
+|—— utils.py # 数据集结果保存等工具函数
+|—— prompt/ # FewCLUE 各数据集的 prompt 定义文件
+```
```
Comment (wawltor) [inline examples/few_shot/p-tuning/README.md:23]: example的升级安装model zoo的标准来，把train.py 改成 run_train.py

File: examples/few_shot/p-tuning/README.md [hunk 2]
```diff
@@ -1,113 +1,82 @@
...
+
 ###  数据准备
-基于 FewCLUE 数据集进行实验只需要  1 行代码，这部分代码在 `ptuning.py` 脚本中
+基于 FewCLUE 数据集进行实验只需要  1 行代码，这部分代码在 `data.py` 脚本中
```
Comment (wawltor) [inline examples/few_shot/p-tuning/README.md]: example的升级安装model zoo的标准来，把train.py 改成 run_train.py

File: examples/few_shot/p-tuning/README.md [hunk 3]
```diff
@@ -1,113 +1,82 @@
...
 
 ```
 from paddlenlp.datasets import load_dataset
 
 # 通过指定 "fewclue" 和数据集名字 name="tnews" 即可一键加载 FewCLUE 中的 tnews 数据集
 train_ds, dev_ds, public_test_ds = load_dataset("fewclue", name="tnews", splits=("train_0", "dev_0", "test_public"))
-````
```
Comment (wawltor) [inline examples/few_shot/p-tuning/README.md:26]: 这里大致说明一下fewclue数据集、tnews数据集

File: examples/few_shot/p-tuning/README.md [hunk 4]
```diff
@@ -1,113 +1,82 @@
...
-    ptuning.py \
-    --task_name "my_task"
-    --device gpu \
-    --p_embedding_num 1 \
-    --save_dir "checkpoints" \
-    --batch_size 32 \
-    --learning_rate 5E-5 \
-    --epochs 10 \
-    --max_seq_length 512
-```
```
Comment (wawltor) [inline examples/few_shot/p-tuning/README.md:83]: 这里有个TODO 后续加上python的deploy + serving的部署

File: examples/few_shot/p-tuning/train.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+    if data_args.do_test and public_test_ds is not None:
+        test_compute_metrics = partial(compute_metrics,
+                                       labels=test_labels,
+                                       verbalizer=verbalizer)
+        trainer.compute_metrics = test_compute_metrics
+        test_ret = trainer.predict(public_test_ds)
+        trainer.log_metrics("test", test_ret.metrics)
+
+    # Predict.
+    if training_args.do_predict and test_ds is not None:
```
Comment (wawltor) [inline examples/few_shot/p-tuning/train.py]: 后续还是需要增加一个显式的predict.py 来 做部署

File: examples/few_shot/p-tuning/utils.py [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+        if isinstance(label_words, list):
+            label_words = {k: k for k in label_words}
+        args.label_words = label_words
+        return args
+
+
+def maskedlm_wrapper(verbalizer):
+
+    def process_outputs(self,
+                        outputs: Tensor,
```
Comment (wawltor) [inline examples/few_shot/p-tuning/utils.py]: 这里的函数命名这块看看要不要去掉Tensor，或者其他的函数保持统一

File: examples/few_shot/p-tuning/utils.py [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+    label_probs = np.max(preds, axis=1)
+    pseudo_data = []
+    for index, example in enumerate(data_ds):
+        example["labels"] = labels[label_preds[index]]
+        example["prob"] = str(label_probs[index])
+        pseudo_data.append(example)
+    save_data(pseudo_data, save_path)
+
+
+def save_fewclue_prediction(save_path, task_name, label_preds, verbalizer,
```
Comment (wawltor) [inline examples/few_shot/p-tuning/utils.py]: 这个函数主要是为了fewclue的打榜的结果吗？

File: examples/few_shot/pet/README.md [hunk 8]
```diff
@@ -1,75 +1,81 @@
...
 
-[PET](https://arxiv.org/abs/2001.07676) (Exploiting Cloze Questions for Few Shot Text Classification and Natural Language Inference)  提出将输入示例转换为完形填空式短语，以帮助语言模型理解给定的任务
 
-## 代码结构及说明
+[Exploiting Cloze Questions for Few Shot Text Classification and Natural Language Inference](https://arxiv.org/abs/2001.07676)
+
+
+## 摘要
+
+自然语言处理任务可以通过给预训练模型提供“任务描述”等方式来进行无监督学习，但效果一般低于有监督训练。而 Pattern-Exploiting Training (PET) 是一种半监督方法，通过将输入转换为完形填空形式的短语来帮助语言模型理解任务。然后用这些短语来给无标注数据打软标签。最后在得到的标注数据集上用有监督方法进行训练。在小样本设置下，PET 在部分任务上远超有监督学习和强半监督学习方法。
```
Comment (wawltor) [inline examples/few_shot/pet/README.md]: 这里加上一个PET的建模的示意图

File: examples/few_shot/pet/README.md [hunk 9]
```diff
@@ -1,75 +1,81 @@
...
+|—— utils.py # 数据集构造、数据增强等工具函数
+|—— prompt/  # FewCLUE 各数据集的 prompt 定义文件
 ```
 
-
 ## 基于 FewCLUE 进行 PET 实验
 PaddleNLP 内置了 FewCLUE 数据集，可以直接用来进行 PET 策略训练、评估、预测，并生成 FewCLUE 榜单的提交结果，参与 FewCLUE 竞赛。
 
 ###  数据准备
-基于 FewCLUE 数据集进行实验只需要  1 行代码，这部分代码在 `pet.py` 脚本中
```
Comment (wawltor) [inline examples/few_shot/pet/README.md:19]: 同上

File: examples/few_shot/pet/README.md [hunk 10]
```diff
@@ -1,75 +1,81 @@
...
+- `do_label`: 是否使用训练后的模型给无标签数据标注伪标签
+- `do_test`: 是否在公开测试集上评估模型效果
+- `model_name_or_path`: 预训练模型名，默认为 `ernie-1.0-large-zh-cw`
+- `use_rdrop`: 是否使用对比学习策略 R-Drop
+- `alpha_rdrop`: R-Drop 损失值权重，默认为 0.5
+- `dropout`: 预训练模型的 dropout 参数值，用于 R-Drop 策略中参数配置
+- `export_type`: 模型导出格式，默认为 `paddle`，动态图转静态图
+- 更多配置参考 [Trainer 参数文档](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/docs/trainer.md#trainingarguments-%E5%8F%82%E6%95%B0%E4%BB%8B%E7%BB%8D) 和 [PromptTrainer 参数文档](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/docs/advanced_guide/prompt.md#prompttrainer%E5%8F%82%E6%95%B0%E5%88%97%E8%A1%A8)
 
-### 模型导出
```
Comment (wawltor) [inline examples/few_shot/pet/README.md:66]: TODO 后续加上相关的模型部署逻辑

File: examples/few_shot/p-tuning/train.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        attention_probs_dropout_prob=model_args.dropout)
+
+    # Define template for preprocess and verbalizer for postprocess.
+    template = SoftTemplate(data_args.prompt, tokenizer,
+                            training_args.max_seq_length, model)
+    logger.info("Using template: {}".format(template.prompt))
+
+    verbalizer = ManualVerbalizer(data_args.label_words, tokenizer)
+    verbalizer.mask_aggregate_type = "mean"
+    verbalizer = maskedlm_wrapper(verbalizer)
```
Comment (wawltor) [inline examples/few_shot/p-tuning/train.py]: 这里为什么每个算法都需要单独wrap一下了？ 竞品这块的实现方式是怎么样的？

File: examples/few_shot/p-tuning/utils.py (hunk 7 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/utils.py]: 是的，按照FewCLUE提交格式保存结果。

File: examples/few_shot/p-tuning/train.py (hunk 11 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/train.py]: 这里是之前打的补丁，新版本发布后直接调用 `MaskedLMVerbalizer` 即可。这部分逻辑会删掉。

File: examples/few_shot/p-tuning/utils.py (hunk 6 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/utils.py]: 这个函数是打的补丁，已删除。

File: examples/few_shot/p-tuning/README.md (hunk 1 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/README.md:23]: 已改为run_train.py。

File: examples/few_shot/p-tuning/README.md (hunk 3 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/README.md:26]: 已增加fewclue数据集的简介。

File: examples/few_shot/pet/README.md (hunk 8 above)
Comment (LemonNoel) [reply examples/few_shot/pet/README.md]: 已增加PET模型图。

File: examples/few_shot/pet/README.md (hunk 9 above)
Comment (LemonNoel) [reply examples/few_shot/pet/README.md:19]: 同上

File: examples/few_shot/p-tuning/README.md (hunk 4 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/README.md:83]: 已在文档中增加该项，后续由张斌负责。

File: examples/few_shot/p-tuning/train.py (hunk 5 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/train.py]: 同上

File: examples/few_shot/pet/README.md (hunk 10 above)
Comment (LemonNoel) [reply examples/few_shot/pet/README.md:66]: 已在文档中增加该项，后续由张斌负责。

**Review by wawltor — DISMISSED** [review]
LGTM

File: examples/few_shot/p-tuning/README.md (hunk 2 above)
Comment (LemonNoel) [reply examples/few_shot/p-tuning/README.md]: 当前训练入口均为 run_train.py

**Review by wawltor — APPROVED** [review]
LGTM
